### PR TITLE
G-API GMatP

### PR DESCRIPTION
**Pull request by rgarnov**
* Added GMatP type
* Changed planar tests to operate on GMatP

### CODE REVIEW

File: modules/gapi/include/opencv2/gapi/cpu/gcpukernel.hpp [hunk 1]
```diff
@@ -188,6 +192,14 @@ template<> struct get_out<cv::GMat>
         return {r};
     }
 };
+template<> struct get_out<cv::GMatP>
+{
+    static tracked_cv_mat get(GCPUContext &ctx, int idx)
+    {
+        auto& r = ctx.outMatR(idx);
+        return {r};
```
Comment (dmatveev) [inline modules/gapi/include/opencv2/gapi/cpu/gcpukernel.hpp]: y not oneliner?

File: modules/gapi/include/opencv2/gapi/gmat.hpp [hunk 2]
```diff
@@ -46,6 +46,19 @@ class GAPI_EXPORTS GMat
...
+public:
+    GMatP();                                 // Empty constructor
+    GMatP(const GNode &n, std::size_t out);  // Operation result constructor
+
+    GOrigin& priv();                        // Internal use only
+    const GOrigin& priv()  const;           // Internal use only
+
+private:
+    std::shared_ptr<GOrigin> m_priv;
+};
```
Comment (dmatveev) [inline modules/gapi/include/opencv2/gapi/gmat.hpp]: Could this be just subclassed from regular `cv::GMat` (with all fields reused)}

File: modules/gapi/include/opencv2/gapi/gtype_traits.hpp [hunk 3]
```diff
@@ -47,6 +48,11 @@ namespace detail
         static constexpr const ArgKind kind = ArgKind::GMAT;
         static constexpr const GShape shape = GShape::GMAT;
     };
+    template<>           struct GTypeTraits<cv::GMatP>
+    {
+        static constexpr const ArgKind kind = ArgKind::GMATP;
```
Comment (dmatveev) [inline modules/gapi/include/opencv2/gapi/gtype_traits.hpp]: Why does it need a special Kind enum?

Comment (rgarnov) [reply modules/gapi/include/opencv2/gapi/gtype_traits.hpp]: It's required by GArg->GProtoArg conversion

File: modules/gapi/include/opencv2/gapi/gmat.hpp (hunk 2 above)
Comment (rgarnov) [reply modules/gapi/include/opencv2/gapi/gmat.hpp]: Yes, will fix

File: modules/gapi/include/opencv2/gapi/cpu/gcpukernel.hpp (hunk 1 above)
Comment (rgarnov) [reply modules/gapi/include/opencv2/gapi/cpu/gcpukernel.hpp]: Reworked

**Comment by rgarnov** [comment]
The build is red due to GProtoArg variant extension with new type

**Pull request merged**